compiler/optimizer/LocalOpts.cpp

```diff
@@ -6900,27 +6889,34 @@ int32_t TR_InvariantArgumentPreexistence::perform()
 
          const char **argInfo = peekInfo->_args;
          int32_t *lenInfo = peekInfo->_lengths;
-         int32_t index = 0;
+         index = 0;
          for (TR::ParameterSymbol *p = parms.getFirst(); p != NULL; p = parms.getNext(), index++)
             {
-            const char *sig = NULL;
-            if (argInfo && argInfo[index])
+            ParmInfo &parmInfo = _parmInfo[index];
+            TR_OpaqueClassBlock *clazzFromMethod = parmInfo.getClass();
+            if (argInfo && argInfo[index] && parmInfo.isInvariant() && !parmInfo.classIsCurrentlyFinal() && clazzFromMethod)
                {
-               sig = argInfo[index];
+               const char *sig = argInfo[index];
                int32_t len = lenInfo[index];
                TR_ASSERT(((*sig == 'L') || (*sig == '[')), "non address argument cannot be of fixed ref type");
 
-               _parmInfo[index].setSymbol(p);
-
                TR_OpaqueClassBlock *clazz = fe()->getClassFromSignature(sig, len, feMethod);
-               if (clazz)
+               if (!clazz || clazz == clazzFromMethod)
+                  continue;
+
+               // Peeking arg info is not compatible with the method, looking at dead path, bail out
+               if (fe()->isInstanceOf(clazz, clazzFromMethod, true, true, true) != TR_yes)
```

Comment:
would a trace message be useful here? Seems like knowing which parm we failed on would be helpful in debugging or did I miss it printing somewhere else?

Reply:
Trace added.